Migrate to the new Quick settings menu and more

pakaoraki opened the pull request:
Gnome 43 introduced a new quick settings toggle menu and I propose to rewrite the extension to follow this evolution.

![screenshot](https://user-images.githubusercontent.com/83842133/196008308-b8659115-d525-4501-944b-faff480b6b82.png)

Changelog:
-Drop Panel menu and migrate to QuickSettingsMenu.
-Scroll up/down the indicator icon to enable/disable caffeine instead of clicking it (click open the toggle menu now).
-Update 'Show status icon' option with one new state 'only when active'.
-Upgrade notifications to the latest Gnome OSD notification.
-Update README and screenshots.

Pros/cons: using a toggle menu instead of panel icon cost one more click, but it's still possible to scroll directly on the status icon so it still quickly accessible.

My repo: [https://github.com/pakaoraki/gnome-shell-extension-caffeine](https://github.com/pakaoraki/gnome-shell-extension-caffeine)

I use this extension since a while now, thanks for everybody who works on it. Feel free to merge or cherry-pick any code that you want to use in Caffeine. I'm please for any feedback or modification on this pull request.

Cheers

pakaoraki commented:
If you like, I also made some news changes:

- New Preferences window with Libadwaita + GTK4:
![screenshot-prefs](https://user-images.githubusercontent.com/83842133/200126540-f8773627-6ae9-4450-a556-36325b02a973.png)

- There is a new timer options in QuickToggle Menu.
![screenshot-timer-on](https://user-images.githubusercontent.com/83842133/200126559-05d7c771-128d-4c03-813a-1faf474cd6ff.png)

- The icon been switched to the left in status indicator:
![screenshot](https://user-images.githubusercontent.com/83842133/200126585-c6afe929-8f3c-4385-99da-01e6596fd24d.png)

Sources: [https://github.com/pakaoraki/gnome-shell-extension-caffeine](https://github.com/pakaoraki/gnome-shell-extension-caffeine)

I can make an additional Pull request if you want.

Cheers

agam778 commented:
Pretty good work, to be honest.

![image](https://user-images.githubusercontent.com/68941022/200256729-979101b3-44b3-4d80-85e3-61d95f3b2d77.png)
How about maintaining this extension?

pakaoraki commented:
News changes since last time:
- Add an 'allow blank screen' option.
- Add an option to move the icon position in status indicator menu.
- Fix an annoying bug with recent GTK4/flatpak app that failed to remove inhibitor when triggered Caffeine.
- Improve behavior of how pause/resume Night light mode.
- Improve notifications and clean code.

"How about maintaining this extension?"
-> Honestly, I will be really please to do it. If users of Caffeine like my modifications and if the creator/actual maintainer approve it, I'm in.

eonpatapon commented:
Thanks @pakaoraki, great work!

I will take some time to review this. If other users can test this and give some feedback it would be great.

pakaoraki commented:
> Thanks @pakaoraki, great work!
> 
> I will take some time to review this. If other users can test this and give some feedback it would be great.

Sound great !

felso commented:
I don't see the extension after doing:

```shell
gh pr checkout 216
make build
make install```

It might be that my environment isn't setup properly. There weren't any errors though, and I see it in: ~/.local/share/gnome-shell/extensions

felso commented:
I really like it. The whole reason I was visiting this project on GitHub was that I was thinking this extension needed to be updated for Gnome 43. Very nice.

Maybe you should bump the version to 43.

calebmeyer commented:
I would love to test this, I just have to figure out how. 

I'm getting an error at the extension manager using the fork: 
![image](https://user-images.githubusercontent.com/3342777/202926215-1b4a9dd9-4831-4d96-b39a-cf29237a9c72.png)

pakaoraki commented:
@calebmeyer
I suggest you remove the files you have copied and follow this step to make a clean install :
`git clone https://github.com/pakaoraki/gnome-shell-extension-caffeine.git`
`cd gnome-shell-extension-caffeine`
`make build`
`make install`

Then log out and log in.
I use Fedora 37 as well, it should be working fine.

stuarthayhurst commented:
Came here planning to add this myself, but since you've beaten me too it, I gave it a test :)
All working on my GNOME 43 Debian Sid system, thanks!

stuarthayhurst commented on README.md:
```diff
@@ -1,31 +1,70 @@
-    - `gnome-shell-3.32-3.34`: `3.32` -> `3.34`
-    - `gnome-shell-3.10-3.30`: `3.10` -> `3.30`
-    - `gnome-shell-before-3.10`: `3.4` -> `3.8`
+[<img src="https://github.com/pakaoraki/gnome-shell-extension-caffeine/raw/master/ressources/get_it_on_gnome_extensions.png" height="100" align="right">](https://extensions.gnome.org/extension/517/caffeine/)
```
Should this link be to "https://github.com/eonpatapon/..." ?

pakaoraki replied:
Thanks for your feedback.
folder name fixed.

"Should this link be to "https://github.com/eonpatapon/..." ?"
I did it on purpose as the resource does not exist yet on the main repo. That's a small change to do after merging.

codingdave commented:
I just tested it on Manjaro and am very happy as everything seems to work flawlessly. Thanks for the good work!

eject7455 commented:
> Gnome 43 introduced a new quick settings toggle menu and I propose to rewrite the extension to follow this evolution.
> 
> ![screenshot](https://user-images.githubusercontent.com/83842133/196008308-b8659115-d525-4501-944b-faff480b6b82.png)
> 
> Changelog: -Drop Panel menu and migrate to QuickSettingsMenu. -Scroll up/down the indicator icon to enable/disable caffeine instead of clicking it (click open the toggle menu now). -Update 'Show status icon' option with one new state 'only when active'. -Upgrade notifications to the latest Gnome OSD notification. -Update README and screenshots.
> 
> Pros/cons: using a toggle menu instead of panel icon cost one more click, but it's still possible to scroll directly on the status icon so it still quickly accessible.
> 
> My repo: https://github.com/pakaoraki/gnome-shell-extension-caffeine
> 
> I use this extension since a while now, thanks for everybody who works on it. Feel free to merge or cherry-pick any code that you want to use in Caffeine. I'm please for any feedback or modification on this pull request.
> 
> Cheers

@pakaoraki thanks for this version. It seems much more useful. I found it because standard caffeine didn't do what I wanted. I was looking to only stop auto-suspend and still allow the screens to go off and lock (to save power) because suspend was crashing the pc upon resume when running a VM. I wanted the ability to auto-suspend when not running a VM at least. 
Unfortunately, you asked for feedback and your extension didn't work as expected. You have in the extension preferences the ability to have screen black allowed, but when I adjusted it from Never, the screens still did not blank. It's still functioning all or nothing like regular caffeine. Also you have an option to auto turn on when you launch an app. So I assigned that to virtual machine manager. However the caffeine icon will flicker on and off after I open virt manager. The only option that would successfully automatically toggle it on and off  fully is to use the option for when apps go full screen. 
Sorry, not a programmer so I can't offer any suggestions.
I was also wondering if it was possible for your extension or any others like this to have broken the ability of my computer to lock altogether? I uninstalled through Extension Manager. If not, it must be one of the recent Fedora updates. I can no longer lock the screen. Lock option is missing from quick menu, super+L does not work and after time passes it only goes blank and doesn't lock.

felso commented:
@pakaoraki I'm experiencing a bug where the screen won't blank even when Caffeine is not active and there is no full-screen application. I'm not sure what's triggering it. I will continue to test.

felso commented:
@pakaoraki I'm experiencing a bug where the screen won't blank even when Caffeine is not active and there is no full-screen application. I'm not sure what's triggering it. I will continue to test.

pakaoraki commented:
Thanks for your feedback

@eject7455

> "You have in the extension preferences the ability to have screen blank allowed, but when I adjusted it from Never, the screens still did not blank. It's still functioning all or nothing like regular caffeine"

Option "Allow screen blank"
I just test it again on Fedora 37 / Ubuntu 22.10 and this option works as expected for me: when "always" is selected screen can dim and be locked but automatic suspend is still off. When "never" is selected, caffeine inhibit completely the desktop, no dim, no lock and no suspend.

What's your distro/version ? Can you reproduce this bug and can you try it again after a fresh reboot ? 
Also can you check your power settings maybe :
![image](https://user-images.githubusercontent.com/83842133/209125243-99d58961-2e78-4e9f-9c07-bc9eac98627e.png)

> "Also you have an option to auto turn on when you launch an app. So I assigned that to virtual machine manager. However the caffeine icon will flicker on and off after I open virt manager. The only option that would successfully automatically toggle it on and off fully is to use the option for when apps go full screen."

Do you have the very last version I made installed ? if not reinstall the extension please.
I installed virtual manager and try to reproduce this behavior but it work as expected on my machine: when virt is started, Caffeine automatically turn on and I got any flickering. It is flickering only with virt or have you try other apps ? 
If someone else read this, can you reproduce this bug ?
I will do some more test.

@felso 

> "I'm experiencing a bug where the screen won't blank even when Caffeine is not active and there is no full-screen application. I'm not sure what's triggering it. I will continue to test."

thanks for the test: is it still going after disabling/enabling Caffeine ? It can be some other apps that inhibit the desktop (Firefox, video player,...). When it's happening, can you try to close every apps and see if still going ?

Also, when you change some Caffeine options when it's active, toggle off Caffeine and on again because it won't update the previous inhibited flag when changing "Allow blank" option  for example.

What's distro/version do you use ?

eject7455 commented:
About to go on vacation, but when I get back I will double check the settings and I will try to trigger it with a different app as well. I built from source just a week or two ago with instructions you gave someone else so it should be up to date. I might have been changing options when it was active. I'll look at that took. I am running Fedora 37. Sending updates in about two weeks. Thanks!!
------- Original Message -------
On Thursday, December 22nd, 2022 at 7:07 AM, pakaoraki ***@***.***> wrote:

> Thanks for your feedback
>
> ***@***.***(https://github.com/eject7455)
>
>> "You have in the extension preferences the ability to have screen blank allowed, but when I adjusted it from Never, the screens still did not blank. It's still functioning all or nothing like regular caffeine"
>
> Option "Allow screen blank"
> I just test it again on Fedora 37 / Ubuntu 22.10 and this option works as expected for me: when "always" is selected screen can dim and be locked but automatic suspend is still off. When "never" is selected, caffeine inhibit completely the desktop, no dim, no lock and no suspend.
>
> What's your distro/version ? Can you reproduce this bug and can you try it again after a fresh reboot ?
> Also can you check your power settings maybe :
> [image](https://user-images.githubusercontent.com/83842133/209125243-99d58961-2e78-4e9f-9c07-bc9eac98627e.png)
>
>> "Also you have an option to auto turn on when you launch an app. So I assigned that to virtual machine manager. However the caffeine icon will flicker on and off after I open virt manager. The only option that would successfully automatically toggle it on and off fully is to use the option for when apps go full screen."
>
> Do you have the very last version I made installed ? if not reinstall the extension please.
> I installed virtual manager and try to reproduce this behavior but it work as expected on my machine: when virt is started, Caffeine automatically turn on and I got any flickering. It is flickering only with virt or have you try other apps ?
> If someone else read this, can you reproduce this bug ?
> I will do some more test.
>
> ***@***.***(https://github.com/felso)
>
>> "I'm experiencing a bug where the screen won't blank even when Caffeine is not active and there is no full-screen application. I'm not sure what's triggering it. I will continue to test."
>
> thanks for the test: is it still going after disabling/enabling Caffeine ? It can be some other apps that inhibit the desktop (Firefox, video player,...). When it's happening, can you try to close every apps and see if still going ?
>
> Also, when you change some Caffeine options when it's active, toggle off Caffeine and on again because it won't update the previous inhibited flag when changing "Allow blank" option for example.
>
> What's distro/version do you use ?
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/eonpatapon/gnome-shell-extension-caffeine/pull/216#issuecomment-1362763294), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/A44R6ABMAT6UEGWH5GS446LWOQ773ANCNFSM6AAAAAARGCF73E).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

pakaoraki commented:
After deeper test, I can confirm a problem with triggering apps. So when you go to some menu or dialogs in triggered apps, it continuing switch on/off Caffeine. It's happening with virtual Manager and most others apps. I never notice this behavior before, maybe due to some recent updates ?

Anyway, I think it's related to the way this extension detect windows events, it's little bit tricky. I spent some time on a similar issue before with parasites events emitted when opening app. I will try to debug this when I can.

stuarthayhurst commented:
@pakaoraki @eonpatapon Just checking in, is anything else required for this to be reviewed / merged? I've been running it for nearly a month with 0 issues, in case it's testing that's needed.

Thanks :)

pakaoraki commented:
Hi, after some issues reported with apps triggering, I came up with a new version. I partly rewrite the way apps trigger Caffeine with three new mods:
-"Running": Apps trigger Caffeine when running (Default).
-"Focus": Apps trigger Caffeine only when being focused.
-"Active Workspace": Apps trigger Caffeine only when they are on the active workspace.

![image](https://user-images.githubusercontent.com/83842133/211204463-d860134a-bb61-4fbe-9de3-543b1e0b2cea.png)

The toggle button is now synchronized with Caffeine state: before it was only active when toggled by user.
I also optimize/clean some code as well.

I don't plan to add new features in the future, as it's now a bit overkill maybe, but I will be listening if people really need something that make sense. I've been testing this version for a week mostly on Fedora 37 with no issue, especially with triggering apps that I was focus on.

Let me know if you notice something wrong.

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -332,53 +774,188 @@ class Caffeine extends PanelMenu.Button {
+                this.addInhibit(appId); // Inhibit app
             }
+            
+            // Add 200 ms delay before unblock state signal
```
Can you explain why you need to add a timer here ?

pakaoraki replied:
This is a workaround to solve an issue with unwanted multiple signals when an app start.

You were using "windows-changed" to detect windows event that call "_appWindowsChanged" function:
```
 windowsChangedId: app.connect('windows-changed',
                    this._appWindowsChanged.bind(this)),
```
([https://gjs-docs.gnome.org/shell01~0.1/shell.app#signal-windows-changed](https://gjs-docs.gnome.org/shell01~0.1/shell.app#signal-windows-changed)) 

Now I use a different method using "app-state-changed":
```
this._appStateChangedSignalId =
	this._appSystem.connect('app-state-changed', this._appStateChanged.bind(this)); 
```
([https://gjs-docs.gnome.org/shell01~0.1/shell.appsystem#signal-app-state-changed](https://gjs-docs.gnome.org/shell01~0.1/shell.appsystem#signal-app-state-changed))
    
In both case, for some reasons I got multiple events for few millisecond after the app started. 

For example, with "Black Box" app when starting (from logs):
```
App: com.raggesilver.BlackBox.desktop, state change: 1 (STARTING)
App: com.raggesilver.BlackBox.desktop, state change: 2 (RUNNING)
App: com.raggesilver.BlackBox.desktop, state change: 0 (STOPPED)
App: com.raggesilver.BlackBox.desktop, state change: 2 (RUNNING)
App: com.raggesilver.BlackBox.desktop, state change: 0 (STOPPED)
...
```
(test with `log("App: " + appId + ", state change: " + appState);`  in a dedicated testing function called by "windows-changed" or "app-state-changed")

With the timer, I choose to ignore for 200 ms the next following events, then I switch on again listener. So this way I keep only the first valid one.
Back in November, I got 10 signals at start for the same event mixing RUNNING, STOPPED and STARTING state. Sometimes it's creating multiple inhibited flags and the extension can get stuck so I introduce this solution (see original commit: [https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/bab83b0f1d578bd821ef1a6cd4f8864a9df909f6](https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/bab83b0f1d578bd821ef1a6cd4f8864a9df909f6) and [https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/76b3fc094603f4189eb31e49e4b94319f19b061a](https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/76b3fc094603f4189eb31e49e4b94319f19b061a)).

I'm not sure if it's a bug from the shell/gjs. I redo some test recently and it seems that's the situation is a bit better: I got less signals but still 4 consecutive RUNNING/STOOPED signals for "Black box" terminal app when starting.

eonpatapon replied:
Thanks for the detailled answer! got it

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -141,43 +275,113 @@ class Caffeine extends PanelMenu.Button {
-        if (this._settings.get_boolean(USER_ENABLED_KEY) && this._settings.get_boolean(RESTORE_KEY))
+        
+        // Store the inhibition requests until processed
+        this._inhibition_added_fifo=[];
+        this._inhibition_removed_fifo=[];
```
To keep the style consistent:
```suggestion
        this._inhibitionAddedFifo = [];
        this._inhibitionRemovedFifo = [];
```

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -141,43 +275,113 @@ class Caffeine extends PanelMenu.Button {
+        this._indicator.gicon = this._icon_desactived;
+
+        // Manage night light and allow blank screen
+        this._night_light = false;
+        this._allow_blank= false;
```
```suggestion
        this._nightLight = false;
        this._allowBlank= false;
```

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -141,43 +275,113 @@ class Caffeine extends PanelMenu.Button {
+        this.add_child(this._timerLabel);
 
+        // Icons
+        this._icon_actived = Gio.icon_new_for_string(`${Me.path}/icons/${EnabledIcon}.svg`);;
+        this._icon_desactived = Gio.icon_new_for_string(`${Me.path}/icons/${DisabledIcon}.svg`);
```
```suggestion
        this._iconActived = Gio.icon_new_for_string(`${Me.path}/icons/${EnabledIcon}.svg`);;
        this._iconDesactived = Gio.icon_new_for_string(`${Me.path}/icons/${DisabledIcon}.svg`);
```

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -191,67 +395,251 @@ class Caffeine extends PanelMenu.Button {
+            appData.isToggled = false;
+            this._appInhibitedData.set(appId, appData);            
+        }
+        this._inhibition_removed_fifo.push(appId);
```
Shouldn't this be done before calling `this._sessionManager.UninhibitRemote` ?

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -191,67 +395,251 @@ class Caffeine extends PanelMenu.Button {
+                this._appInhibitedData.set(appId, data);
             }
         );
+        this._inhibition_added_fifo.push(appId);
```
Shouldn't this be done in the `InhibitRemote` callback ? So that we push the appID only if the call to `InhibitRemote` succeeds

pakaoraki replied:
Indeed, make sense.

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -260,70 +648,124 @@ class Caffeine extends PanelMenu.Button {
+                    this._indicator.gicon = this._icon_desactived;
+                            
+                    // Shell OSD notifications
+                    if (this._settings.get_boolean(SHOW_NOTIFICATIONS_KEY))
+                    this._sendOSDNotification(false);
```
NIT: indent
```suggestion
                    if (this._settings.get_boolean(SHOW_NOTIFICATIONS_KEY)) {
                        this._sendOSDNotification(false);
                    }
```

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -260,70 +648,124 @@ class Caffeine extends PanelMenu.Button {
+        if(requested_id){
+            let appData = this._appInhibitedData.get(requested_id);        
+            if (appData){
+                if (requested_id === 'user')
+                       this._saveUserState(false);
```
NIT: indent 
```suggestion
                if (requested_id === 'user') {
                    this._saveUserState(false);
                }
```

pakaoraki replied:
I changed it. 

```
                if (requested_id === 'user')
                       this._saveUserState(false);
```
Is this syntax better to avoid ? 
I thought it was just a more compact way to use when having only one short line of code.

eonpatapon replied:
It's fine. Saves some lines but sometimes it think it's harder to read

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -260,70 +648,124 @@ class Caffeine extends PanelMenu.Button {
-                if (this._settings.get_boolean(SHOW_NOTIFICATIONS_KEY))
-                    this._sendNotification('disabled');
+        // Get the first removed request 
+        let requested_id = this._inhibition_removed_fifo.shift();
```
I suggest to keep the name appId so we know what it is

pakaoraki replied:
I changed it.

eonpatapon commented on caffeine@patapon.info/extension.js:
```diff
@@ -191,67 +395,251 @@ class Caffeine extends PanelMenu.Button {
-            this._manageNightLight('disabled');
-        }
+            this._manageNightLight(false, false);
+        }        
```
There's quite a lot of trailing spaces in the file, they should be removed

pakaoraki replied:
Most of the time, I try to get rid of it. I'll correct this.

pakaoraki commented:
Thanks for this review, I updated the code.

stuarthayhurst commented:
Does the `gnome-shell-40-43` branch need creating from `86116f3f7d2c7959abad4b433db9ff9844d93d70`?

eonpatapon commented:
@stuarthayhurst will do

eonpatapon commented:
@stuarthayhurst @pakaoraki I've uploaded the new version on extensions.gnome.org

stuarthayhurst commented:
Strange, I got the update earlier, and then another 'update' reverted to version 42. Was it pulled?

eonpatapon commented:
There's a few things to fix:

>     You should remove the timeouts on destroy (line 400, 892, 901 and 951 extension.js):
>     https://gjs.guide/extensions/review-guidelines/review-guidelines.html#remove-main-loop-sources

stuarthayhurst commented:
Is this just a matter of saving the ID of the timeout, setting it to `null` when the timeout is handled, then in `disable()`, disconnecting from any of these IDs that aren't `null`?

eonpatapon commented:
Yes, like `this._timePrint`, `this._timeOut`

stuarthayhurst commented:
Ok, I'll work on a PR for this. Is there any reason `setTimeout()` is used, rather than `GLib.timeout_add`?

stuarthayhurst commented:
Also, is `Mainloop` required (`imports.mainloop`), or can `GLib` be used instead?

stuarthayhurst commented:
Assuming I'm good to use `GLib`, this should be sorted by #223

pakaoraki commented:
Hi, thanks a lot @stuarthayhurst. I didn't really think about `GLib.timeout_add`, I just needed a way to add a delay. That's looking really good now.
